flocker/common/algebraic.py

```diff
@@ -0,0 +1,49 @@
+# -*- test-case-name: flocker.common.test.test_algebraic -*-
+# Copyright Hybrid Logic Ltd.  See LICENSE file for details.
+
+"""
+Invariants for algebraic data types.
+"""
+
+from pyrsistent import PClass, field
+
+__all__ = ["StateInvariant"]
+
+
+class StateInvariant(PClass):
```

Comment:
It's a bit confusing to use the term "state" (here & elsewhere), because it overlaps with the abstract concept of programming state, and also with states in our state machine, which is where I assume this is derived from. It would be confusing in a different way to use the term "constructor", which is more common in Haskell. 

Looking at https://doc.rust-lang.org/book/enums.html for inspiration, `TaggedUnionInvariant`, with `states` renamed to `tags` and `state` to `tag` might be more clear. 

Alternatively, `EnumInvariant`, either with the "tag" nomenclature, or `variant`.